jena-base/src/main/java/org/apache/jena/atlas/lib/tuple/Tuple.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.atlas.lib.tuple;
+
+import java.util.List ;
+import java.util.function.Consumer ;
+import java.util.stream.Stream ;
+
+import org.apache.jena.atlas.lib.ArrayUtils ;
+
+/** A Tuple is the same class of item */
+public interface Tuple<X> extends Iterable<X> {
+    /** Get the i'th element, for i in the range 0 to len()-1 
+     * @throws IndexOutOfBoundsException for i out of range 
+     */
+    public X get(int i) ;
+
+    /** length : elements are 0 to len()-1 */ 
+    public int len() ;
+
+    /** Convert to a List */
+    public default List<X> asList() {
+        return new TupleList<>(this) ;
+    }
+
+    /** stream */
+    public default Stream<X> stream() { 
+        return asList().stream() ;
+    }
+
+    /** forEach */
+    @Override
+    public default void forEach(Consumer<? super X> action) { 
+        asList().forEach(action) ;
+    }
+
+    /** Copy the Tuple into the array */ 
+    public default void copyInto(X[] array) {
+        copyInto(array, 0, len());
+    }
+
+    /** Copy the Tuple into the array */ 
+    public default void copyInto(X[] array, int start) {
+        copyInto(array, start, len());
+    }
+
+    /** Copy the Tuple into the array */ 
+    public void copyInto(X[] array, int start, int length) ;
```

Comment:
Isn't there a really obvious `default` here? Maybe it would be good to have a comment here explaining why impls need to do their own thing here...?